spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/observation/ObservedAppDeployer.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2018-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.deployer.spi.app.observation;
+
+import java.time.Duration;
+
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.deployer.spi.app.AppDeployer;
+import org.springframework.cloud.deployer.spi.app.AppScaleRequest;
+import org.springframework.cloud.deployer.spi.app.AppStatus;
+import org.springframework.cloud.deployer.spi.app.DeploymentState;
+import org.springframework.cloud.deployer.spi.core.AppDeploymentRequest;
+import org.springframework.cloud.deployer.spi.core.RuntimeEnvironmentInfo;
+import org.springframework.lang.Nullable;
+
+/**
+ * Observed representation of an {@link AppDeployer}.
+ *
+ * @author Marcin Grzejszczak
+ */
+public class ObservedAppDeployer implements AppDeployer {
+
+	private static final Log log = LogFactory.getLog(ObservedAppDeployer.class);
+
+	private static final DefaultAppDeployerObservationConvention DEFAULT_CONVENTION = new DefaultAppDeployerObservationConvention();
+	private final AppDeployer delegate;
+
+	private final ObservationRegistry observationRegistry;
+
+	@Nullable
+	private final AppDeployerObservationConvention customConvention;
+
+	private final Long pollDelay;
+
+	public ObservedAppDeployer(ObservationRegistry observationRegistry, @Nullable AppDeployerObservationConvention customConvention, AppDeployer delegate, Long pollDelay) {
+		this.observationRegistry = observationRegistry;
+		this.delegate = delegate;
+		this.customConvention = customConvention;
+		this.pollDelay = pollDelay;
+	}
+
+	public ObservedAppDeployer(ObservationRegistry observationRegistry, AppDeployer delegate, Long pollDelay) {
+		this(observationRegistry, null, delegate, pollDelay);
+	}
+
+	@Override
+	public String deploy(AppDeploymentRequest request) {
```

Comment:
We also have a `TaskLauncher.launch` and `TaskLauncher.cancel` for ephemeral applications (Jobs).